diff -ur grub2/util/grub-install.in grub2.new/util/grub-install.in
--- grub2/util/grub-install.in	2011-06-28 13:04:15 +0000
+++ grub2.new/util/grub-install.in	2011-09-15 10:55:32 +0000
@@ -627,9 +627,9 @@
     fi
 
     # Try to make this image bootable using the EFI Boot Manager, if available.
-    efibootmgr="`which efibootmgr`"
+    Efibootmgr="`which efibootmgr`"
     if test "$removable" = no && test -n "$efi_distributor" && \
-	test -n "$efibootmgr"; then
+	test -n "$Efibootmgr"; then
         # On Linux, we need the efivars kernel modules.
 	case "$host_os" in
 	    linux*)
@@ -637,9 +637,9 @@
 	esac
 
         # Delete old entries from the same distributor.
-	for bootnum in `efibootmgr | grep '^Boot[0-9]' | \
+	for bootnum in `$Efibootmgr | grep '^Boot[0-9]' | \
 	    fgrep -i " $efi_distributor" | cut -b5-8`; do
-	    efibootmgr $efi_quiet -b "$bootnum" -B
+	    $Efibootmgr $efi_quiet -b "$bootnum" -B
 	done
 
         # Add a new entry for the image we just created.  efibootmgr needs to be
@@ -654,7 +654,7 @@
 	else
 	    efidir_disk="$(echo "$clean_devmap" | grep "^$(echo "$efidir_drive" | sed 's/,[^)]*//')" | cut -f2)"
 	    efidir_part="$(echo "$efidir_drive" | sed 's/^([^,]*,[^0-9]*//; s/[^0-9].*//')"
-	    efibootmgr $efi_quiet -c -d "$efidir_disk" -p "$efidir_part" -w \
+	    $Efibootmgr $efi_quiet -c -d "$efidir_disk" -p "$efidir_part" -w \
 		-L "$bootloader_id" -l "\\EFI\\$efi_distributor\\$efi_file"
 	fi
     fi
