diff -Naur grub-2.00-orig/util/getroot.c grub-2.00/util/getroot.c
--- grub-2.00-orig/util/getroot.c	2012-06-22 18:42:07.000000000 +0300
+++ grub-2.00/util/getroot.c	2012-10-29 19:53:38.031403149 +0200
@@ -1025,6 +1025,60 @@
 
 #endif /* __CYGWIN__ */
 
+static char *
+__workaround_evms (char *os_dev)
+{
+  if (os_dev)
+    {
+      char *dest;
+      int i;
+
+#ifdef DEBUG
+      fprintf(stderr, "__workaround_evms: '%s' -> ", os_dev);
+#endif
+
+      if (dest = strstr (os_dev, "/evms/lvm2/"))
+        {
+          char *c;
+
+          strcpy (dest, "/mapper/");
+          strcat (dest, dest + 11);
+          c = strchr (dest + 8, '/');
+          if (c)
+            *c = '-';
+        }
+      else if (dest = strstr (os_dev, "/evms/"))
+        i = 5;
+      else if (dest = strstr (os_dev,"/mapper/sd"))
+        i = 7;
+      else if (dest = strstr (os_dev,"/mapper/vd"))
+        i = 7;
+      else
+        i = 0;
+
+      if (i)
+        strcpy (dest, dest + i);
+
+#ifdef DEBUG
+      fprintf(stderr, "'%s'\n", os_dev);
+#endif
+    }
+  return os_dev;
+}
+
+static char *
+workaround_evms (char **os_dev)
+{
+  if (os_dev)
+    {
+      char **ods;
+
+      for (ods = os_dev; *ods; ods++)
+        __workaround_evms (*ods);
+    }
+  return os_dev;
+}
+
 char **
 grub_guess_root_devices (const char *dir)
 {
@@ -1035,7 +1085,7 @@
 
 #ifdef __linux__
   if (!os_dev)
-    os_dev = grub_find_root_devices_from_mountinfo (dir, NULL);
+    os_dev = workaround_evms (grub_find_root_devices_from_mountinfo (dir, NULL));
 #endif /* __linux__ */
 
 #if !defined (__MINGW32__) && !defined (__CYGWIN__)
@@ -1098,6 +1148,7 @@
 
   /* This might be truly slow, but is there any better way?  */
   os_dev[0] = grub_find_device ("/dev", dev);
+  __workaround_evms (os_dev[0]);
 #endif
   if (!os_dev[0])
     {
@@ -2597,7 +2648,7 @@
       grub_dev = grub_util_biosdisk_get_grub_dev (os_dev);
     }
 
-  return grub_dev;
+  return __workaround_evms (grub_dev);
 }
 
 const char *

diff -Naur grub-2.00-orig/util/grub-install.in grub-2.00/util/grub-install.in
--- grub-2.00-orig/util/grub-install.in	2012-06-27 00:31:03.000000000 +0300
+++ grub-2.00/util/grub-install.in	2012-10-26 19:10:43.301911972 +0300
@@ -564,13 +564,14 @@
 
 # Write device to a variable so we don't have to traverse /dev every time.
 grub_device="`"$grub_probe" --device-map="${device_map}" --target=device "${grubdir}"`" || exit 1
+grub_device_fs=`echo $grub_device | sed 's|/dev/mapper/\([sv]d\)|/dev/\1|g' `
 
 if ! test -f "${grubdir}"/grubenv; then
     "$grub_editenv" "${grubdir}"/grubenv create
 fi
 
 # Create the core image. First, auto-detect the filesystem module.
-fs_module="`echo "${grub_device}" | xargs "$grub_probe" --device-map="${device_map}" --target=fs --device `"
+fs_module="`echo "${grub_device_fs}" | xargs "$grub_probe" --device-map="${device_map}" --target=fs --device `"
 if test "x$fs_module" = x ; then
     gettext_printf "Auto-detection of a filesystem of %s failed.\n" "${grub_device}" 1>&2
     gettext "Try with --recheck." 1>&2
@@ -650,7 +651,7 @@
         # generic method (used on coreboot and ata mod)
 	uuid=
 	if [ x"$force_file_id" != xy ]; then
-            uuid="`echo "${grub_device}" | xargs "$grub_probe" --device-map="${device_map}" --target=fs_uuid --device`"
+            uuid="`echo "${grub_device_fs}" | xargs "$grub_probe" --device-map="${device_map}" --target=fs_uuid --device`"
 	fi
 
 	if [ x"$disk_module" != x ] && [ x"$disk_module" != xbiosdisk ]; then
