diff -Naur grub-2.00-orig/util/grub-install.in grub-2.00/util/grub-install.in
--- grub-2.00-orig/util/grub-install.in	2012-06-27 00:31:03.000000000 +0300
+++ grub-2.00/util/grub-install.in	2012-10-26 19:53:02.016915133 +0300
@@ -738,6 +738,8 @@
 # Perform the grub_modinfo_platform-dependent install
 if [ "${grub_modinfo_target_cpu}-${grub_modinfo_platform}" = "i386-pc" ] || [ "${grub_modinfo_target_cpu}-${grub_modinfo_platform}" = "sparc64-ieee1275" ] ; then
     # Now perform the installation.
+    sync
+    sleep 2
     "$grub_setup" ${allow_floppy} ${setup_verbose} ${setup_force} --directory="${grubdir}/${grub_modinfo_target_cpu}-$grub_modinfo_platform" \
 	--device-map="${device_map}" "${install_device}" || exit 1
 elif [ "${grub_modinfo_target_cpu}-${grub_modinfo_platform}" = "i386-ieee1275" ] || [ "${grub_modinfo_target_cpu}-${grub_modinfo_platform}" = "powerpc-ieee1275" ]; then
