diff -ur grub2/util/grub-mkconfig.in grub2.new/util/grub-mkconfig.in
--- grub2/util/grub-mkconfig.in	2010-03-06 20:51:37 +0000
+++ grub2.new/util/grub-mkconfig.in	2010-03-06 21:36:28 +0000
@@ -130,8 +130,8 @@
 # choosing Hurd filesystem module.
 GRUB_FS="`${grub_probe} --target=fs / 2> /dev/null || echo unknown`"
 
-if test -f ${sysconfdir}/default/grub ; then
-  . ${sysconfdir}/default/grub
+if test -f ${sysconfdir}/sysconfig/grub2 ; then
+  . ${sysconfdir}/sysconfig/grub2
 fi
 
 # XXX: should this be deprecated at some point?
