diff -ur grub2/kern/emu/getroot.c grub2.new/kern/emu/getroot.c
--- grub2/kern/emu/getroot.c	2010-08-02 19:44:02 +0000
+++ grub2.new/kern/emu/getroot.c	2010-08-04 12:24:23 +0000
@@ -535,6 +535,16 @@
 
   /* This might be truly slow, but is there any better way?  */
   os_dev = find_root_device ("/dev", st.st_dev);
+
+  char *dest;
+  if ((os_dev)&&(dest=strstr(os_dev,"/evms/"))&&(!strstr(os_dev,"/evms/lvm2/")))
+    {
+        char *tail=(char *)malloc(strlen(strstr(os_dev,"/evms/")+5)+1);
+        strncpy(tail,strstr(os_dev,"/evms/")+5,strlen(strstr(os_dev,"/evms/")+5)+1);
+        strcpy(dest,tail);
+        free(tail);
+    }
+
 #endif
 #endif /* !__GNU__ */
 
@@ -670,6 +680,7 @@
 grub_util_get_grub_dev (const char *os_dev)
 {
   char *grub_dev = NULL;
+  char *dest = NULL;
 
   switch (grub_util_get_dev_abstraction (os_dev))
     {
@@ -786,6 +797,14 @@
       grub_dev = grub_util_biosdisk_get_grub_dev (os_dev);
     }
 
+  if ((grub_dev)&&(dest=strstr(grub_dev,"/evms/"))&&(!strstr(grub_dev,"/evms/lvm2/")))
+    {
+        char *tail=(char *)malloc(strlen(strstr(grub_dev,"/evms/")+5)+1);
+        strncpy(tail,strstr(grub_dev,"/evms/")+5,strlen(strstr(grub_dev,"/evms/")+5)+1);
+        strcpy(dest,tail);
+        free(tail);
+    }
+
   return grub_dev;
 }
 
