diff --git a/grub2/util/grub.d/10_linux.in b/grub2/util/grub.d/10_linux.in
index 14b85c7..374353a 100644
--- a/grub2/util/grub.d/10_linux.in
+++ b/grub2/util/grub.d/10_linux.in
@@ -28,11 +28,10 @@ export TEXTDOMAINDIR=@localedir@
 
 CLASS="--class gnu-linux --class gnu --class os"
 
-if [ "x${GRUB_DISTRIBUTOR}" = "x" ] ; then
-  OS=GNU/Linux
+if [ -r /etc/altlinux-release ]; then
+  OS=`cat /etc/altlinux-release | sed 's/ (.*$//g'`
 else
-  OS="${GRUB_DISTRIBUTOR} GNU/Linux"
-  CLASS="--class $(echo ${GRUB_DISTRIBUTOR} | tr '[A-Z]' '[a-z]' | cut -d' ' -f1) ${CLASS}"
+  OS=ALTLinux
 fi
 
 # loop-AES arranges things so that /dev/loop/X can be our root device, but
@@ -57,12 +56,19 @@ linux_entry ()
   version="$2"
   recovery="$3"
   args="$4"
+  default="$5"
+  title="$(gettext "%s, with Linux %s")"
+  if ${default} ; then
+    title="`echo $title | sed 's|, .*$||g'`"
+  fi
   if ${recovery} ; then
-    title="$(gettext_quoted "%s, with Linux %s (recovery mode)")"
+     title="$title (failsafe mode)"
+  fi
+  if ${default} ; then
+     printf "menuentry \"${title}\" ${CLASS} {\n" "${os}"
   else
-    title="$(gettext_quoted "%s, with Linux %s")"
+     printf "menuentry \"${title}\" ${CLASS} {\n" "${os}" "${version}"
   fi
-  printf "menuentry '${title}' ${CLASS} {\n" "${os}" "${version}"
   save_default_entry | sed -e "s/^/\t/"
 
   # Use ELILO's generic "efifb" when it's known to be available.
@@ -102,24 +108,21 @@ EOF
 EOF
 }
 
-list=`for i in /boot/vmlinu[zx]-* /vmlinu[zx]-* ; do
-        if grub_file_is_not_garbage "$i" ; then echo -n "$i " ; fi
-      done`
-prepare_boot_cache=
-
-while [ "x$list" != "x" ] ; do
-  linux=`version_find_latest $list`
+ls -c /boot/vmlinuz /boot/vmlinu[xz]-* | while read linux ; do
   echo "Found linux image: $linux" >&2
   basename=`basename $linux`
   dirname=`dirname $linux`
+  initrd=
+  default=false
+  linux_root_device_thisversion="${LINUX_ROOT_DEVICE}"
   rel_dirname=`make_system_path_relative_to_its_root $dirname`
+  if [ x$linux != x/boot/vmlinuz ]; then
   version=`echo $basename | sed -e "s,^[^0-9]*-,,g"`
+  flavour=`echo $basename | sed -e "s,^vmlinuz-,,g"`
   alt_version=`echo $version | sed -e "s,\.old$,,g"`
-  linux_root_device_thisversion="${LINUX_ROOT_DEVICE}"
-
-  initrd=
   for i in "initrd.img-${version}" "initrd-${version}.img" \
 	   "initrd-${version}" "initramfs-${version}.img" \
+	   "initrd-${flavour}.img" \
 	   "initrd.img-${alt_version}" "initrd-${alt_version}.img" \
 	   "initrd-${alt_version}" "initramfs-${alt_version}.img"; do
     if test -e "${dirname}/${i}" ; then
@@ -127,6 +130,12 @@ while [ "x$list" != "x" ] ; do
       break
     fi
   done
+  else
+    if [ -e /boot/initrd.img ]; then 
+      initrd="initrd.img"
+    fi
+    default=true
+  fi
   if test -n "${initrd}" ; then
     echo "Found initrd image: ${dirname}/${initrd}" >&2
   else
@@ -135,11 +144,10 @@ while [ "x$list" != "x" ] ; do
   fi
 
   linux_entry "${OS}" "${version}" false \
-      "${GRUB_CMDLINE_LINUX} ${GRUB_CMDLINE_LINUX_DEFAULT}"
+      "${GRUB_CMDLINE_LINUX_DEFAULT}" $default
   if [ "x${GRUB_DISABLE_LINUX_RECOVERY}" != "xtrue" ]; then
     linux_entry "${OS}" "${version}" true \
-	"single ${GRUB_CMDLINE_LINUX}"
+	"${GRUB_CMDLINE_LINUX}" $default
   fi
 
-  list=`echo $list | tr ' ' '\n' | grep -vx $linux | tr '\n' ' '`
 done
