diff --git a/grub2/util/grub-probe.c b/grub2/util/grub-probe.c
index ba2fe4c..9e4e78f 100644
--- a/grub2/util/grub-probe.c
+++ b/grub2/util/grub-probe.c
@@ -112,6 +112,7 @@ probe (const char *path, char *device_name)
   char *grub_path = NULL;
   char *filebuf_via_grub = NULL, *filebuf_via_sys = NULL;
   grub_device_t dev = NULL;
+  char *dest = NULL;
   grub_fs_t fs;
 
   if (path == NULL)
@@ -127,6 +128,14 @@ probe (const char *path, char *device_name)
   else
     device_name = grub_guess_root_device (path);
 
+  if (dest=strstr(device_name,"/evms/"))
+    {
+        char *tail=(char *)malloc(strlen(strstr(device_name,"/evms/")+5)+1);
+        strncpy(tail,strstr(device_name,"/evms/")+5,strlen(strstr(device_name,"/evms/")+5)+1);
+        strcpy(dest,tail);
+        free(tail);
+    }
+
   if (! device_name)
     grub_util_error ("cannot find a device for %s (is /dev mounted?)", path);
 
