--- grub-core/loader/efi/linux.c.orig	2019-02-12 14:22:10.273735569 +0000
+++ grub-core/loader/efi/linux.c	2019-02-12 14:26:41.939419278 +0000
@@ -57,10 +57,19 @@
       return 1;
     }
 
-  grub_dprintf ("secureboot", "Kernel signature verification failed (0x%lx)\n",
+  grub_dprintf ("secureboot", "Kernel signature verification failed (0x%lx)\n"
+#ifdef RELAX_KERNEL_SIGNATURE_CHECK
+		"WARNING! Override failure, continue boot!\n",
+#else
+		,
+#endif
 		(unsigned long) status);
 
+#ifdef RELAX_KERNEL_SIGNATURE_CHECK
+  return 1;
+#else
   return -1;
+#endif
 }
 
 typedef void (*handover_func) (void *, grub_efi_system_table_t *, void *);
