diff --git a/grub/grub-core/osdep/unix/getroot.c b/grub/grub-core/osdep/unix/getroot.c
index 4bf37b0..2964dcd 100644
--- a/grub/grub-core/osdep/unix/getroot.c
+++ b/grub/grub-core/osdep/unix/getroot.c
@@ -433,7 +433,8 @@ grub_find_device (const char *dir, dev_t dev)
 	      ent->d_name[1] == 'm' &&
 	      ent->d_name[2] == '-' &&
 	      ent->d_name[3] >= '0' &&
-	      ent->d_name[3] <= '9')
+	      ent->d_name[3] <= '9' &&
+	      ent->d_name[4] == '\0')
 	    continue;
 #endif
 
diff --git a/grub/util/grub.d/10_linux.in b/grub/util/grub.d/10_linux.in
index de9044c..6a96222 100644
--- a/grub/util/grub.d/10_linux.in
+++ b/grub/util/grub.d/10_linux.in
@@ -47,7 +47,7 @@ esac
 # and mounting btrfs requires user space scanning, so force UUID in this case.
 if [ "x${GRUB_DEVICE_UUID}" = "x" ] || [ "x${GRUB_DISABLE_LINUX_UUID}" = "xtrue" ] \
     || ! blkid -U "${GRUB_DEVICE_UUID}" > /dev/null \
-    || ( test -e "${GRUB_DEVICE}" && uses_abstraction "${GRUB_DEVICE}" lvm ); then
+    || ( test -e "${GRUB_DEVICE}" && uses_abstraction "${GRUB_DEVICE}" lvm && ! uses_abstraction "${GRUB_DEVICE}" luks ); then
   LINUX_ROOT_DEVICE=${GRUB_DEVICE}
 else
   LINUX_ROOT_DEVICE=UUID=${GRUB_DEVICE_UUID}
