diff --git a/grub2/util/grub-install.in b/grub2/util/grub-install.in
index ff8bea8..a7854a6 100644
--- a/grub2/util/grub-install.in
+++ b/grub2/util/grub-install.in
@@ -581,6 +581,8 @@ fi
 # Perform the platform-dependent install
 if [ "${target_cpu}-${platform}" = "i386-pc" ] || [ "${target_cpu}-${platform}" = "sparc64-ieee1275" ] ; then
     # Now perform the installation.
+    sync
+    sleep 2s
     "$grub_setup" ${allow_floppy} ${setup_verbose} ${setup_force} --directory="${grubdir}" \
 	--device-map="${device_map}" "${install_device}" || exit 1
 elif [ "${target_cpu}-${platform}" = "i386-ieee1275" ] || [ "${target_cpu}-${platform}" = "powerpc-ieee1275" ]; then
